src/Services/QueuedJobService.php

```diff
@@ -334,6 +300,98 @@ public function isAtMaxJobs()
         return false;
     }
 
+    /**
+     * Using a job signature, returns the JobDescriptor ID and whether the
+     * job descriptor is new or existing
+     *
+     * @param string $signature
+     * @param QueuedJob $job
+     * @param QueuedJobDescriptor $jobDescriptor
+     * @param null|string $startAfter
+     * @return int|null
+     */
+    protected function findOrMakeJobDescriptorFromSignature($signature, $job, $jobDescriptor, $startAfter)
+    {
+        // Start a transaction which will hold until we have a lock on this signature.
+        return DB::get_conn()->withTransaction(function () use ($signature, $job, $jobDescriptor, $startAfter) {
+            $query = 'SELECT "ID" FROM "QueuedJobDescriptor" WHERE "Signature" = ? FOR UPDATE';
+
+            // Retrieve first record
+            $result = DB::prepared_query($query, [$signature]);
+
+            if ($result === null) {
+                throw new Exception('Failed to execute query to retrieve job signature');
+            }
+
+            $ID = $result->value();
+
+            // If the record does not exist
+            if (!$ID) {
+                // use this to populate custom data columns before job is queued
+                // note: you can pass arbitrary data to your job and then move it to job descriptor
+                // this is useful if you need some data that needs to be exposed as a separate
+                // DB column as opposed to serialised data
+                $this->extend('updateJobDescriptorBeforeQueued', $jobDescriptor, $job);
+
+                // copy data
+                $this->copyJobToDescriptor($job, $jobDescriptor);
+
+                // Write the record
+                $jobDescriptorID = $jobDescriptor->write();
+
+                $this->startJob($jobDescriptor, $startAfter);
+            } else {
+                $jobDescriptorID = $ID;
+            }
+        });
+    }
+
+    /**
+     * @param QueuedJob $job
+     * @param string $signature
+     * @param null $startAfter
+     * @param null $userId
+     * @param null $queueName
+     * @return QueuedJobDescriptor
+     */
+    protected function createJobDescriptor(
+        QueuedJob $job,
+        $signature,
+        $startAfter = null,
+        $userId = null,
```

Comment:
```suggestion
        $userID = null,
```
Convention is to capitilise the d in ID

Also update the references in this function to `$userID`